feature/discover-interface

cohen-seth opened the pull request:
This PR adds the changes to multiple scripts which allow the observation-inventory-utils to work on Discover (NASA). AWS and HPSS functionality are preserved despite these edits:

(1) Two new files have been added: obs_inv_utils_discover.csh, discover_interface.py. 
(2) Changes were made to 5 scripts: nceplibs_bufr_cmd_handler.py, obs_inv_queries.py, obs_storage_platforms.py, search_engine.py 

Files for testing the changes above have also been added: 
(3) Four test yaml files have also been added: test_discover.sh, test_aws_amsua_1.yaml, test_aws_amsua_2.yaml, test_discover_amsua_1.yaml, test_discover_amsua_2.yaml.  
(4) A .sh file, test_discover.sh, that runs the appropriate obs_inv_cli commands for all the new yamls for faster testing on discover.

cohen-seth committed:
feature/discover-interface edits

cohen-seth committed:
feature/discover-interface edits

cohen-seth committed:
feature/discover-interface nceplibs_bufr_cmd_handler edits and discover test file

cohen-seth committed:
test_discover wrkdir edits

cohen-seth committed:
test_discover.sh

jrknezha commented on src/obs_inv_utils/nceplibs_bufr_cmd_handler.py:
```diff
@@ -23,6 +23,8 @@
 from obs_inv_utils.subprocess_cmd_handler import SubprocessCmd
 from obs_inv_utils import nceplibs_cmds as nc_cmds
 
+import yaml
```
this import should be grouped above with the others, under uuid

jrknezha commented on src/obs_inv_utils/nceplibs_bufr_cmd_handler.py:
```diff
@@ -120,29 +128,62 @@ def get_bufr_file_meta(self, cmd_type):
 
-        work_dir = os.path.join(self.meta_config.work_dir, temp_uuid)
+        # Because s3 bucket bufr files require download and new work_dir and discover files do not they need to be handled differently. 
+        if bucket == 'noaa-reanalyses-pds': 
```
This is not the appropriate way to do this. We should not be hard coding in the bucket, especially knowing we need to be able to support the private s3 buckets as well in the future. 

If we have to hard code something, it should be the check for discover as that is a much more specific use case.

jrknezha commented on src/obs_inv_utils/nceplibs_bufr_cmd_handler.py:
```diff
@@ -120,29 +128,62 @@ def get_bufr_file_meta(self, cmd_type):
-        work_dir = os.path.join(self.meta_config.work_dir, temp_uuid)
+        # Because s3 bucket bufr files require download and new work_dir and discover files do not they need to be handled differently. 
+        if bucket == 'noaa-reanalyses-pds': 
+            print(f'Running get_bufr_file_meta for NOAA S3 Clean Bucket (aws_s3) {self.s3_bucket}')
```
We should not be referring to it as Clean Bucket, but NNJA. However, this should be much more generalized for any possible s3 bucket

jrknezha commented on src/obs_inv_utils/nceplibs_bufr_cmd_handler.py:
```diff
@@ -120,29 +128,62 @@ def get_bufr_file_meta(self, cmd_type):
+                    os.remove( saved_filename )
+                    shutil.rmtree( work_dir )
+              
+        elif self.s3_bucket == None or self.s3_bucket == '':
```
We should be using a different way to check directly for Discover such as platform

jrknezha replied:
Adding a note for myself to look into: how is HPSS called and would that not be prevented with this elif statement?

jrknezha commented on src/tests/yaml_configs/test_aws_amsua_2.yaml:
```diff
@@ -0,0 +1,11 @@
+s3_bucket: noaa-reanalyses-pds
+s3_prefix: observations/reanalysis/gps/gpsro/%Y/%m/bufr/
```
the file name says amsua but it's pointing to gps data? need to make sure the file name of the file and the bufr files is consistent with the s3_prefix

jrknezha commented on src/tests/test_discover.sh:
```diff
@@ -0,0 +1,12 @@
+python src/obs_inv_utils/obs_inv_cli.py get-obs-inventory -c src/tests/yaml_configs/test_aws_amsua_1.yaml ; 
+python src/obs_inv_utils/obs_inv_cli.py get-obs-count-meta-sinv -c src/tests/yaml_configs/test_discover_amsua_2.yaml ;
+
+echo "test complete."
```
this is not actually a test but a chance to run some files, it won't actually check and confirm things are working. Since we are already using pytest for other testing, it would be appropriate to write this into a python file using pytest and assert statements for confirmation that the code is working appropriately

jrknezha commented on src/obs_inv_utils/discover_interface.py:
```diff
@@ -0,0 +1,293 @@
+    ],
+)
+
+HpssTarballContents = namedtuple(
```
is this actually getting used in this file or was it just there for reference making the discover ones? if just reference they can be removed now

jrknezha reviewed (changes requested):
This is a big PR so this is just a first pass review, once we get through some of the code parts I'll make sure to run everything as part of my final review. For now, there's a few areas that I've commented on that need to be changed to be more robust / follow existing patterns / software best practices.

cohen-seth committed:
Addressing Issues from first PR

A change in the configuration of the sinv and cmpbqm input yamls. A 'platform' field was added to these yamls as a way the tool can differentiate between platforms.

Corrected some typos and changed some comments to be more accurate.

cohen-seth committed:
Merge branch 'NOAA-PSL:develop' into develop
